addons/udes_stock/models/stock_picking_batch.py

```diff
@@ -55,6 +84,87 @@ def _compute_priority(self):
             if new_priority != batch.priority:
                 batch.priority = new_priority
 
+    @api.multi
+    @api.constrains('user_id')
+    def _compute_state(self):
+        """ Compute the state of a batch post confirm
+            waiting    : Aleast some picks are not ready
+            ready      : All picks are in ready state (assigned)
+            in_progess : All picks are ready and a user has been assigned
+            done       : All picks are complete (in state done or cancel)
+
+            the other two states are draft and cancel are manual
+            to tranistions from or to the respective state.
+        """
+        Picking = self.env['stock.picking']
+
+        for batch in self:
+
+            if batch.state in ['draft', 'done', 'cancel']:
+                # Can't do anything with them don't bother trying.
+                continue
+
+            if not batch.picking_ids:
+                batch.state = 'done'
+                continue
+
+            ready_picks = Picking.browse()
+            other_picks = Picking.browse()
+            complete_picks = Picking.browse()
+
+            # Collect picks into groups
+            for pick in batch.picking_ids:
+
+                if pick.state == 'assigned':
+                    ready_picks |= pick
+
+                elif pick.state in ['done', 'cancel']:
+                    complete_picks |= pick
+
+                else:
+                    other_picks |= pick
+
+            # Based on groups compute batch state
```

Comment:
As discussed, I suggest to be explicit about the current state, transitions and assumptions (i.e. dispatch the computation based on the current state).